drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+{
+	int ret;
+
+	ret = virtio_vdpa_virtq_doorbell_relay_disable(priv, index);
+	if (ret) {
+		 DRV_LOG(ERR, "Virtq doorbell relay failed");
+		 return ret;
+	}
+
+	virtio_pci_dev_queue_del(priv->vpdev, index);
+
+	priv->vrings[index]->notifier_state = VIRTIO_VDPA_NOTIFIER_STATE_DISABLED;
+
+	ret = virtio_pci_dev_interrupts_disable(priv->vpdev, index+1);
+	if (ret) {
+		 DRV_LOG(ERR, "Virtq %d interrupt disabel failed", index);
+		 return ret;
+	}
+
+	priv->vrings[index]->enable = 0;
+	return 0;
+}
+
+static int
+virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int index)
+{
+	int ret;
+	int vid;
+	struct rte_vhost_vring vq;
+	struct virtio_pci_dev_vring_info vring_info;
+	uint64_t gpa;
+
+	vid = priv->vid;
+
+	ret = rte_vhost_get_vhost_vring(vid, index, &vq);
+	if (ret)
+		return ret;
+
+	ret =virtio_pci_dev_interrupts_enable(priv->vpdev, vq.callfd, index+1);
+	if (ret) {
+		 DRV_LOG(ERR, "Virtq interrupt enable failed");
+		 return ret;
+	}
+
+	gpa = virtio_vdpa_hva_to_gpa(vid, (uint64_t)(uintptr_t)vq.desc);
+	if (gpa == 0) {
+		DRV_LOG(ERR, "Virtio vdpa dev %s fail to get GPA for descriptor ring %d.",
+						priv->vdev->device->name, index);
+		return -EINVAL;
+	}
+	DRV_LOG(DEBUG, "Virtq %d desc addr%"PRIx64, index, gpa);
+	priv->vrings[index]->desc = gpa;
+	vring_info.desc = gpa;
+
+	gpa = virtio_vdpa_hva_to_gpa(vid, (uint64_t)(uintptr_t)vq.avail);
+	if (gpa == 0) {
+		DRV_LOG(ERR, "Fail to get GPA for available ring.");
+		return -EINVAL;
+	}
+	DRV_LOG(DEBUG, "Virtq %d avail addr%"PRIx64, index, gpa);
+	priv->vrings[index]->avail = gpa;
+	vring_info.avail = gpa;
+
+	gpa = virtio_vdpa_hva_to_gpa(vid, (uint64_t)(uintptr_t)vq.used);
+	if (gpa == 0) {
+		DRV_LOG(ERR, "Fail to get GPA for used ring.");
+		return -EINVAL;
+	}
+	DRV_LOG(DEBUG, "Virtq %d used addr%"PRIx64, index, gpa);
+	priv->vrings[index]->used = gpa;
+	vring_info.used = gpa;
+
+	//TO_DO: need to check vq_size not exceed hw limit
+	priv->vrings[index]->size = vq.size;
+	vring_info.size = vq.size;
+
+	DRV_LOG(DEBUG, "Virtq %d nr_entrys:%d",index, vq.size);
+	if (virtio_pci_dev_queue_set(priv->vpdev, index, &vring_info)) {
+			DRV_LOG(ERR, "Setup_queue failed");
+			return -EINVAL;
+	}
+
+	ret = virtio_vdpa_virtq_doorbell_relay_enable(priv, index);
+	if (ret) {
+		 DRV_LOG(ERR, "Virtq doorbell relay failed");
+		 return ret;
+	}
+
+	priv->vrings[index]->enable = 1;
+	virtio_pci_dev_queue_notify(priv->vpdev, index);
+	return 0;
+}
+
+static int
+virtio_vdpa_vring_state_set(int vid, int vring, int state)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+	int ret = 0;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	if (vring >= (int)priv->hw_nr_virtqs) {
+		DRV_LOG(ERR, "Too big vring id: %d.", vring);
+		return -E2BIG;
+	}
+
+	//TO_DO: check if vid set here is sutiable
+	priv->vid = vid;
```

Comment:
Do

yum install -y codespell
git format-patch -2 -o /tmp/patches/
codespell /tmp/patches/*

This will emit all the spelling errors and fix them.